kvserver: untangle shared/external snap streaming logic

tbg opened the pull request:
This made my head hurt. For the full discovery story, go [here].

If a snapshot may possibly reference external/shared SSTs (i.e. ones not local to the sending node), we avoid putting a range deletion into the addressable portion of the state machine snapshot (i.e. the bulk of the snapshot).

Instead, we require that this snapshot is ingested with an accompanying excise span. However, if we later realize that we cannot actually send shared/external SSTs, we were forcibly _disabling_ excise ingestion (even if it was used for all other snapshots). I believe this was written this way before excise became the default, but now it explicitly steered us into a legacy path. This then awkwardly necessitated sending a message to the recipient to please quickly add a range deletion to the main state machine SSTable:

```
// if req.TransitionFromSharedToRegularReplicate received:
if err := msstw.addClearForMVCCSpan(); err != nil {
  return noSnap, errors.Wrap(err, "adding tombstone for last span")
}
```

Then, the sender would stream the addressable state machine data, this time without relying on external/shared SSTs.

CockroachDB has routinely ingested every snapshot on the excise path since 24.1, and falling into the legacy path here causes complexity, namely the aforementioned call to `addClearForMVCCSpan`. We would like to remove non-excise snapshots altogether (see https://github.com/cockroachdb/cockroach/pull/142651). This PR keeps us on the excise path if we were initially on it.

This turns addClearForMVCCSpan into "obviously" dead code. I'll leave it around for now, since it's used in a unit test and I'd like to tackle it separately.

----

There was also the general question on how this transition path worked. Since the ingested SSTs have to have nonoverlapping bounds (and SSTs must be
sorted), this "transition path" only makes sense no MVCC keys had been streamed yet. Looking at
the code, this is far from obvious. It does appear to be true, though, so an
assertion has been added.
It's worth noting that SST references (shared/external) may have been sent, as the visitor may have seen some of them. So we can't remove the transition path entirely. We could, however, if we buffered up the 

----

The resulting story is now much simpler and makes sense to me:

- Snapshots that have any chance of referencing external/shared SSTs use excise, and don't leave that path.
- They may fall back to not using external/shared SSTs. Since we know haven't streamed any keys or values we from the MVCC span yet by the time we discover this (though we may have told the recipient about some SST references), we send the recipient a "transition" message.
- the only effect of receiving this message is clearing out the slice of external SST references.
- after that, we're effectively on the "regular" snapshot path (with excise).

[here]: https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1741356670269679

Epic: CRDB-46488
Release note: none

tbg committed:
kvserver: stay in excise path when transitioning

tbg committed:
kvserver: assert against transition after having sent MVCC keys

Trust but verify.

tbg committed:
kvserver: remove two conditionals

doExcise is true in this case, and also it's not better to silently drop
dependencies of the snapshot and apply it anyway.

tbg committed:
kvserver: tighten an assertion

InternalKeyKind{Delete{,Sized},RangeDelete,RangeKey{Unset,Delete}}
can only occur in shared SSTs.

Update the assertions to reflect this.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142476)
<!-- Reviewable:end -->

jbowens reviewed (commented):
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142476)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv, @RaduBerinde, and @stevendanna)

___
*[`pkg/kv/kvserver/kv_snapshot_strategy.go` line 206 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-OKmV9yoEh2aQ5Y7pG5p:-OKmV9yoEh2aQ5Y7pG5q:b-jtkms7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a813cffc5b23140abcc0cebaac0c1999be1854c/pkg/kv/kvserver/kv_snapshot_strategy.go#L206)):*
> ```Go
> 				}
> 				// TODO(tbg): document why many key kinds are only possible in the
> 				// excise case.
> ```

I believe the exceptional key kinds are only possible under _shared replication_, where the original sstable originated from another store's LSM and may contain the full set of possible key kinds. Both ordinary sender-constructed sstables and external sstables originating from backups can only contain sets and range key sets. They only express what is _in_ the snapshot, not a delta over some previous LSM state or sstable.
___
*[`pkg/kv/kvserver/kv_snapshot_strategy.go` line 276 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-OKmUlys5i87KGQDtyVE:-OKmUlys5i87KGQDtyVF:b1nxy40) ([raw file](https://github.com/cockroachdb/cockroach/blob/663505585ef0b575e18eab017a85bf2a025c1f76/pkg/kv/kvserver/kv_snapshot_strategy.go#L276)):*
> ```Go
> 		}
> 
> 		if len(req.SharedTables) > 0 {
> ```

nit: in both of these we can now eliminate the conditional altogether
___
*[`pkg/ccl/storageccl/engineccl/shared_storage.go` line 40 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-OKlv-XkBjtPfe66dg0j:-OKlv-XkBjtPfe66dg0k:b-fukt65) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa27e5ad471dcab1299a9cc50ca4d0ba815aaca2/pkg/ccl/storageccl/engineccl/shared_storage.go#L40)):*
> ```Go
> //
> // If this method returns pebble.ErrInvalidSkipSharedIteration, only the
> // shared an external visitors may have been invoked. In particular, no
> ```

nit: rogue "an"


<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: extract readOneToBatch

This declutters the Receive method somewhat.

tbg committed:
kvserver: remove assertions

These are already checked by the callers, and the callers do
the correct check: we can have internal range keys only if
we are getting sent shared SSTs.

The association of shared SSTs with skipping the clear for MVCC spans
is accidental and about to disappear.

tbg reviewed (commented):
This is now ready for review. I ran into some issues which are now described at the top of the PR message.
I also tried to scope down this PR, so there are now a few stacked PRs:

- https://github.com/cockroachdb/cockroach/pull/142648 (the "ran into some issues")
- https://github.com/cockroachdb/cockroach/pull/142651 - bakes in the excise=true cluster setting as discussed on Slack.
- https://github.com/cockroachdb/cockroach/pull/142653 - a change that seems sane but that I feel better having scrutinized in a separate commit.

I'm happy to talk through any of these changes in person. This is a precursor to other inflight work that I have, and would appreciate a timely review so that I can compact the context I'm holding in my head for this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-:-OL4C2xOCFTfpO75T210:b-h29yn7)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens, @pav-kv, @RaduBerinde, and @stevendanna)

___
*[`pkg/kv/kvserver/kv_snapshot_strategy.go` line 276 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-OKmUlys5i87KGQDtyVE:-OL-3utu4ziW5v6LMRFH:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/663505585ef0b575e18eab017a85bf2a025c1f76/pkg/kv/kvserver/kv_snapshot_strategy.go#L276)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: in both of these we can now eliminate the conditional altogether
</blockquote></details>

Done. 
___
*[`pkg/kv/kvserver/kv_snapshot_strategy.go` line 206 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-OKmV9yoEh2aQ5Y7pG5p:-OL-3oF9Biq6Ijixo_7X:b-lj51ue) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a813cffc5b23140abcc0cebaac0c1999be1854c/pkg/kv/kvserver/kv_snapshot_strategy.go#L206)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I believe the exceptional key kinds are only possible under _shared replication_, where the original sstable originated from another store's LSM and may contain the full set of possible key kinds. Both ordinary sender-constructed sstables and external sstables originating from backups can only contain sets and range key sets. They only express what is _in_ the snapshot, not a delta over some previous LSM state or sstable.
</blockquote></details>

Thank you, that makes sense. I added a commit that tightens the assertions to this effect.


<!-- Sent from Reviewable.io -->

tbg committed:
kvserver: note that we can remove shared->reg transition

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> The changes make sense to me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142476#-:-OL5xZMtEVFLK6GE_AWv:b-cqvq5i)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @jbowens, @pav-kv, and @stevendanna)



<!-- Sent from Reviewable.io -->
